server/lib/scheduler/Scheduler.ts

```diff
@@ -0,0 +1,110 @@
+import { filter } from "asyncro";
+
+class Scheduler {
+  private fetchAllWorkspaces;
+  private rootParentCache;
+  private schedule;
+
+  public constructor({ fetchAllWorkspaces, rootParentCache, schedule }) {
+    this.fetchAllWorkspaces = fetchAllWorkspaces;
+    this.rootParentCache = rootParentCache;
+    this.schedule = schedule;
+  }
+
+  public async getIdOfCurrentWorkspace(userId) {
+    const assignment = this.schedule.getMostRecentAssignmentForUser(userId);
+    return assignment.getWorkspaceId();
+  }
+
+  public async findNextWorkspace(userId) {
+    // clear cache so we don't use old eligibility info
+    this.rootParentCache.clearRootParentCache();
+    const idsOfWorkspacesThatCouldBeNext = await this.getIdsOfWorkspacesThatCouldBeNext(userId);
+    const assignedWorkspaceId = this.pickWorkspaceIdAtRandom(idsOfWorkspacesThatCouldBeNext);
+    await this.schedule.assignWorkspaceToUser(userId, assignedWorkspaceId);
+  }
+
+  private async getIdsOfWorkspacesThatCouldBeNext(userId) {
```

Comment:
I don't think this needs the param userId, it doesn't seem to use it.

Reply:
Also, why have this function only return the Ids? If it returns the workspaces, it's easy to filter on them, and the name would be shorter here (this seems done in a few places)

Reply:
Also, I don't like the phrasing "ThatCouldBeNext". Maybe "findActionableWorkspaces" or some other term? ("Valid"?)